mcp_foundry_knowledge/data_access_objects/dao.py

```diff
@@ -0,0 +1,517 @@
+
+        Returns:
+            list[str]: A list of index names.
+        """
+        search_results = self.client.list_index_names()
+        results: list[str] = []
+
+        for search_result in search_results:
+            results.append(search_result)
+
+        return results
+
+    def retrieve_index_schemas(self) -> list[MutableMapping[str, Any]]:
+        """
+        Retrieves the full schema definition for each search index.
+
+        Returns:
+            list[SearchIndex]: A list of serialized index schema definitions.
+        """
+        search_results: ItemPaged[SearchIndex] = self.client.list_indexes()
+        results = []
+
+        for search_result in search_results:
+            results.append(search_result.serialize(keep_readonly=True))
+
+        return results
+
+    def retrieve_index_schema(self, index_name: str) -> MutableMapping[str, Any]:
+        """
+        Retrieves the full schema definition for a search index.
+
+        Returns:
+            SearchIndex: A serialized index schema definition.
+        """
+        search_results = self.client.get_index(index_name)
+
+        return search_results.serialize(keep_readonly=True)
+
+    def modify_index(self, index_name: str, updated_index_definition: SearchIndex) -> MutableMapping[str, Any]:
+        """
+        Updates an existing index in the Azure AI Search service.
+
+        Args:
+            index_name (SearchIndex): The name of the index to be updated
+            updated_index_definition (SearchIndex): The full definition of the index.
+
+        Returns:
+            MutableMapping[str, Any]: The serialized response of the created index.
+        """
+
+        logger.debug(f"Updating Index {index_name} with new definition", updated_index_definition)
+
+        updated_index_definition.name = index_name
+        operation_results = self.client.create_or_update_index(updated_index_definition)
+        return operation_results.serialize(keep_readonly=True)
+
+    def create_index(self, index_definition: SearchIndex) -> MutableMapping[str, Any]:
+        """
+        Creates a new index in the Azure AI Search service.
+
+        Args:
+            index_definition (SearchIndex): The full definition of the index to be created.
+
+        Returns:
+            MutableMapping[str, Any]: The serialized response of the created index.
+        """
+        logger.debug("Creating Index ", index_definition)
+        operation_results = self.client.create_index(index_definition)
+        return operation_results.serialize(keep_readonly=True)
+
+    def delete_index(self, index_name: str):
+        """
+        Deletes an existing index from the Azure AI Search service.
+
+        Args:
+            index_name (str): The name of the index to be deleted.
+
+        Returns:
+            None
+        """
+        logger.debug(f"Deleting Index {index_name}")
+        self.client.delete_index(index_name)
+
+class SearchClientDao(SearchBaseDao):
```

Comment:
Not sure I understand the benefits of a Dao class for the search client

Reply:
In the long run my goal is always to ensure that the code is easily maintainable and this is accomplished by separating the concerns. Having the core logic decoupled from the tool implementations makes the code more testable an maintenable. There are some logic for managing documents that I added some extra logic to make it easier for agents so I did not want to implement this directly in the tool. That is my reasoning behind having the DAOs separate from the tool definitions.